storm-core/src/native/worker-launcher/impl/utils/storm_user_info.c

```diff
@@ -0,0 +1,275 @@
+                         uinfo->buf_sz, &pwd);
+    switch(ret) {
+      case 0:
+        if (!pwd) {
+          // Not found.
+          return ENOENT;
+        }
+        // Found.
+        return 0;
+      case EINTR:
+        // EINTR: a signal was handled and this thread was allowed to continue.
+        break;
+      case ERANGE:
+        // ERANGE: the buffer was not big enough.
+        if (uinfo->buf_sz == MAX_USER_BUFFER_SIZE) {
+          // Already tried with the max size.
+          return ENOMEM;
+        }
+        buf_sz = uinfo->buf_sz * 2;
+        if (buf_sz > MAX_USER_BUFFER_SIZE) {
+          buf_sz = MAX_USER_BUFFER_SIZE;
+        }
+        nbuf = realloc(uinfo->buf, buf_sz);
+        if (!nbuf) {
+          return ENOMEM;
+        }
+        uinfo->buf = nbuf;
+        uinfo->buf_sz = buf_sz;
+        break;
+      default:
+        // Lookup failed.
+        return getpwnam_error_translate(ret);
+    }
+  }
+}
+
+static int put_primary_gid_first(struct storm_user_info *uinfo)
+{
+  int i, num_gids = uinfo->num_gids;
+  gid_t first_gid;
+  gid_t gid;
+  gid_t primary = uinfo->pwd.pw_gid;
+
+  if (num_gids < 1) {
+    // There are no gids, but we expected at least one.
+    return EINVAL;
+  }
+  first_gid = uinfo->gids[0];
+  if (first_gid == primary) {
+    // First gid is already the primary.
+    return 0;
+  }
+  for (i = 1; i < num_gids; i++) {
+    gid = uinfo->gids[i];
+    if (gid == primary) {
+      // swap first gid and this gid.
+      uinfo->gids[0] = gid;
+      uinfo->gids[i] = first_gid;
+      return 0;
+    }
+  }
+  // Did not find the primary gid in the list.
+  return EINVAL;
+}
+
+int storm_user_info_getgroups(struct storm_user_info *uinfo)
+{
+  int ret, ngroups;
+  gid_t *ngids;
+
+  if (!uinfo->pwd.pw_name) {
+    // invalid user info
+    return EINVAL;
+  }
+  uinfo->num_gids = 0;
+  if (!uinfo->gids) {
+    uinfo->gids = malloc(sizeof(uinfo->gids[0]) * INITIAL_GIDS_SIZE);
+    if (!uinfo->gids) {
+      return ENOMEM;
+    }
+    uinfo->gids_size = INITIAL_GIDS_SIZE;
+  }
+  ngroups = uinfo->gids_size;
+  ret = getgrouplist(uinfo->pwd.pw_name, uinfo->pwd.pw_gid, 
+                         uinfo->gids, &ngroups);
+  // Return value is different on Linux vs. FreeBSD.  Linux: the number of groups
+  // or -1 on error.  FreeBSD: 0 on success or -1 on error.  Unfortunately, we
+  // can't accept a 0 return on Linux, because buggy implementations have been
+  // observed to return 0 but leave the other out parameters in an indeterminate
+  // state.  This deviates from the man page, but it has been observed in
+  // practice.  See issue HADOOP-10989 for details.
+#ifdef __linux__
+  if (ret > 0) {
+#else
+  if (ret >= 0) {
+#endif
+    uinfo->num_gids = ngroups;
+    ret = put_primary_gid_first(uinfo);
+    if (ret) {
+      return ret;
+    }
+    return 0;
+  } else if (ret != -1) {
+    // Any return code that is not -1 is considered as error.
```

Comment:
nit: change to all "All negative return codes except -1 is considered as error"

Reply:
when it is on linux, we have
```
if (ret > 0) {
} else if (ret != -1) {
    //ret could be zero
}
```
I will avoid causing confusion here. The original statement look fine to me.